pkg/ruler/ruler_test.go

```diff
@@ -1211,14 +1213,17 @@ func TestRuler_NotifySyncRulesAsync_ShouldTriggerRulesSyncingAndCorrectlyHandleT
 	}
 
 	// We expect rule groups to have been sharded between the rulers.
-	actualRuleGroupsCount := 0
-	for _, ruler := range rulers {
-		actualRuleGroups, err := ruler.getLocalRules(userID, AnyRule)
-		require.NoError(t, err)
-		require.NotEmpty(t, actualRuleGroups)
-		actualRuleGroupsCount += len(actualRuleGroups)
-	}
-	assert.Equal(t, numRuleGroups, actualRuleGroupsCount)
+	test.Poll(t, time.Second, numRuleGroups, func() interface{} {
```

Comment:
I don't think this checks sharding because require.NotEmpty(t, actualRuleGroups)  is not replaced with anything. This it can succeed even if only one ruler has all groups.

Reply:
Good point! I slightly modified the suggestion to return a value which would be easier to debug in case of assertion failures: https://github.com/grafana/mimir/pull/5250/commits/59c0f0954c41a42379ceb1ab0a21c7c9df219975